pkg/config/event_watcher.go

```diff
@@ -94,34 +105,43 @@ func LoadEventWatcher(repoRoot string, includes, excludes []string) (*EventWatch
 
 // filterEventWatcherFiles filters the given files based on the given Includes and Excludes.
 // Excludes are prioritized if both Excludes and Includes are given.
-func filterEventWatcherFiles(files []os.FileInfo, includes, excludes []string) ([]os.FileInfo, error) {
-	if len(includes) == 0 && len(excludes) == 0 {
+func filterEventWatcherFiles(files, includePatterns, excludePatterns []string) ([]string, error) {
+	if len(includePatterns) == 0 && len(excludePatterns) == 0 {
 		return files, nil
 	}
 
-	filtered := make([]os.FileInfo, 0, len(files))
-	useWhitelist := len(includes) != 0 && len(excludes) == 0
-	if useWhitelist {
-		whiteList := make(map[string]struct{}, len(includes))
-		for _, i := range includes {
-			whiteList[i] = struct{}{}
-		}
+	filtered := make([]string, 0, len(files))
+	// Use include patterns
+	if len(includePatterns) != 0 && len(excludePatterns) == 0 {
 		for _, f := range files {
-			if _, ok := whiteList[f.Name()]; ok {
-				filtered = append(filtered, f)
+			// TODO: Cache match results so that it doesn't have to compare include/exclude patterns each time.
+			for _, p := range includePatterns {
+				matched, err := path.Match(p, f)
+				if err != nil {
+					return nil, fmt.Errorf("failed to check if file name metches the given include pattern: %w", err)
+				}
+				if matched {
+					filtered = append(filtered, f)
+					break
+				}
 			}
 		}
 		return filtered, nil
 	}
 
-	blackList := make(map[string]struct{}, len(excludes))
-	for _, e := range excludes {
-		blackList[e] = struct{}{}
-	}
+	// Use exclude patterns
+L:
 	for _, f := range files {
-		if _, ok := blackList[f.Name()]; !ok {
-			filtered = append(filtered, f)
+		for _, p := range excludePatterns {
+			matched, err := path.Match(p, f)
+			if err != nil {
+				return nil, fmt.Errorf("failed to check if file name metches the given exclude pattern: %w", err)
+			}
+			if matched {
+				continue L
+			}
 		}
+		filtered = append(filtered, f)
 	}
```

Comment:
How about something like this

``` golang

out := make([]os.FileInfo, 0, len(files))

if len(includes) > 0 {
    for _, f := range files {
          for _, p := range includes {
              if  path.Match(p, f) {
                 out = append(out, f)
              }
          }
    }
} else {
    out = files
}

filtered := out
out := make([]os.FileInfo, 0, len(filtered))

for _, f := range filtered {
    for _, p := range excludes {
              if  !path.Match(p, f) {
                 out = append(out, f)
              }
          }
}

return out
```

Reply:
I got what you mean, but `excludes` are prioritized if both `excludes` and `includes` are given.
So we have to completely divide logic for `len(includePatterns) > 0 && len(excludePatterns) == 0` and  `len(includePatterns) > 0 && len(excludePatterns) > 0`

Reply:
Besides, we have to check the file path doesn't match all of `excludes`. So this part will include all files:
```go
              if  !path.Match(p, f) {
                 out = append(out, f)
              }
```

Reply:
> Besides, we have to check the file path doesn't match all of excludes. So this part will include all files:

Right, so it can be

``` golang
out := make([]os.FileInfo, 0, len(files))

if len(includes) > 0 {
    for _, f := range files {
          for _, p := range includes {
              if  path.Match(p, f) {
                 out = append(out, f)
              }
          }
    }
} else {
    out = files
}

filtered := out
out := make([]os.FileInfo, 0, len(filtered))

for _, f := range filtered {
    var matched bool
    for _, p := range excludes {
              if  path.Match(p, f) {
                  matched = true
                  break
              }
     }
     if !matched {
                 out = append(out, f)
      }
}

return out
```

Reply:
Btw, we can use our `pkg/filematcher` package to check.
It will create only a  single `matcher` instance for checking all files.
https://github.com/pipe-cd/pipe/blob/master/pkg/filematcher/filematcher.go#L61

Reply:
In this case, `includes` will be used. We have to completely ignore `includes` if `excludes` was given.
```go
if len(includes) > 0 {
```

`pkg/filematcher` looks cool! I'll switch to use it!

Reply:
I feel it is ok to specify both of those fields. For example, including a set but exclude some special cases.

``` yaml
includes:
- foo/*.yaml
excludes:
- foo/ignore.yaml
```

Reply:
I'd say it can be confusing. Depending on the context, the meaning of `ecludes` will be different.

Basically `excludes` means that it includes all files other than specified ones.

But, the `excludes` that you suggested means what to delete from the `includes`. To prevent users from such confusion, we settled on prioritizing `excludes` if both were given.

Reply:
I'm truthfully taking the side of making `includes/excludes` deprecated. So why not keeping the spec of `indludes/excludes`, after this, let's discuss a new option `paths`, that allows pattern matches to be inluded.

Reply:
To be honest, I'm more agree with @nghialv idea. In practice, `append all files in a directory then excludes some specified files` is a common usecase. 🤔

Reply:
I am fine to approve this pull request with the current changes.
Just want to discuss this topic.

From my point of view, ignoring `includes` when they were given is more confusing.
Applying `includes` and `excludes` in order makes it easier to understand and implement.

Here is an example from the Apache Maven resource.

> Of course, we can also have both <includes> and <excludes> elements. For example, if we want to include all text files that does not contain the word "test" in their filename.

https://maven.apache.org/plugins/maven-resources-plugin/examples/include-exclude.html